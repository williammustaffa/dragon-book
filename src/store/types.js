const types = {
  // User actions
  USER_LOGIN: "USER_LOGIN",
  USER_LOGIN_SUCCESS: "USER_LOGIN_SUCCESS",
  USER_LOGIN_FAILURE: "USER_LOGIN_FAILURE",
  USER_LOGOUT: "USER_LOGOUT",

  // User actions
  USER_CHECK_SESSION: "USER_CHECK_SESSION",
  USER_CHECK_SESSION_SUCCESS: "USER_CHECK_SESSION_SUCCESS",
  USER_CHECK_SESSION_FAILURE: "USER_CHECK_SESSION_FAILURE",

  // Profile creation
  CREATE_PROFILE: "CREATE_PROFILE",
  CREATE_PROFILE_SUCCESS: "CREATE_PROFILE_SUCCESS",
  CREATE_PROFILE_FAILURE: "CREATE_PROFILE_FAILURE"
};

export default types;