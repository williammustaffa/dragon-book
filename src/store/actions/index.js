// Session
export { default as userLogin } from "./user/userLogin";
export { default as userLoginSuccess } from "./user/userLoginSuccess";
export { default as userLoginFailure } from "./user/userLoginFailure";
export { default as userLogout } from "./user/userLogout";
export { default as userCheckSession } from "./user/userCheckSession";
export { default as userCheckSessionSuccess } from "./user/userCheckSessionSuccess";
export { default as userCheckSessionFailure } from "./user/userCheckSessionFailure";

// Profile
export { default as createProfile } from "./profiles/createProfile";
export { default as createProfileFailure } from "./profiles/createProfileFailure";
