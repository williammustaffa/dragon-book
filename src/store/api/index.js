
import database from "store/api/database";
import * as R from "ramda";
import { v1 as uuidv1 } from 'uuid';
import request from "./request";

const AUTH_TOKEN_KEY = "auth.token";

class API {
  /**
   * Fake API error delay
   * @param {String} value 
   * @param {Int} time 
   */
  static retrieveDataDelayed(value, time = 500) {
    return new Promise(resolve => {
      setTimeout(() => resolve(value), time);
    });
  }

  /**
   * Fake API error delay
   * @param {String} value 
   * @param {Int} time 
   */
  static retrieveErrorDelayed(value, time = 500) {
    return new Promise((resolve, reject) => {
      setTimeout(() => reject({ message: value }), time);
    });
  }

  /**
   * Create token for user
   * @param {*} param
   */
  static tokenize({ id, email }) {
    return btoa(JSON.stringify({ id, email }));
  }

  /**
   * Get user info from token
   */
  static getTokenizedInfo() {
    const token = localStorage.getItem(AUTH_TOKEN_KEY);
  
    return token && JSON.parse(atob(token));
  }

  /**
   * Parse response object
   * @param {*} res 
   */
  static parseReponse(res) {
    return res.data;
  }

  /**
   * Create a user profile
   */
  createProfile = (payload) => {
    const errors = [];

    const existingUserEmail = database.queryAll("profiles", {
      query: { email: payload.email }
    });

    const existingUserUsername = database.queryAll("profiles", {
      query: { username: payload.username }
    });

    if (existingUserEmail.length) {
      errors.push("The email address provided is already in use.");
    }

    if (existingUserUsername.length) {
      errors.push("The username provided is already in use.");
    }

    if (errors.length) {
      return API.retrieveErrorDelayed(errors.join("\r\n"));
    }

    database.insert("profiles", {
      id: uuidv1(),
      email: payload.email,
      username: payload.username,
      firstName: payload.firstName,
      lastName: payload.lastName,
      description: payload.description,
      password: payload.password,
      registrationDate: Date.now(),
    });

    database.commit();

    // Create user and login
    return this.userLogin(payload);
  }

  /**
   * Check user session an then populate data if needed
   */
  userCheckSession = () => {
    const data = API.getTokenizedInfo();

    const results = database.queryAll("profiles", {
      query: { id: data && data.id },
    });

    if (!results.length) {
      return API.retrieveErrorDelayed("No session found.");
    }

    const user = R.clone(results[0]);

    // No need to delay on first load
    return user;
  }

  /**
   * Do user login by given crendentials
   * @param {object} credentials 
   */
  userLogin({ email, password }) {
    const results = database.queryAll("profiles", {
      query: { email, password },
    });

    if (!results.length) {
      return API.retrieveErrorDelayed("You have entered an invalid username or password");
    }

    const user = R.clone(results[0]);

    // Fake a token storage
    localStorage.setItem(AUTH_TOKEN_KEY, API.tokenize(user));

    return API.retrieveDataDelayed(user, 3000);
  }

  /**
   * Clear user session when he loogs out
   */
  userLogout = () => {
    // Clear token
    localStorage.removeItem(AUTH_TOKEN_KEY);

    return API.retrieveDataDelayed(null, 0);
  }

  /**
   * List dragons
   */
  listDragons = () => {
    return request
      .get("/dragon")
      .then(API.parseReponse);
  }

  /**
   * Get dragon details
   * @param {String} id
   */
  getDragon = (id) => {
    return request
      .get(`/dragon/${id}`)
      .then(API.parseReponse);
  }

  /**
   * Get dragon details
   * @param {Object} data
   */
  createDragon = (data) => {
    return request
      .post("/dragon", data)
      .then(API.parseReponse);
  }

  /**
   * Update dragon details
   * @param {Object} data
   */
  updateDragon = (data) => {
    return request
      .put(`/dragon/${data.id}`, data)
      .then(API.parseReponse);
  }

  /**
   * Update dragon details
   * @param {String} data
   */
  deleteDragon = (id) => {
    return request
      .put(`/dragon/${id}`)
      .then(API.parseReponse);
  }
}

// Create a cached connector
let connectorInstance = null;

function getAPIConnector() {
  if (!connectorInstance) {
    connectorInstance = new API();
  }

  return connectorInstance;
}


export { API, getAPIConnector };