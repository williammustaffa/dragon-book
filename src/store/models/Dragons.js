import Dragon from "./Dragon";

class Dragons {
  constructor(payload = {}) {
    this.items = payload
      .map(dragon => new Dragon(dragon));
  }

  /**
   * Initial reducer state
   */
  static state = {
    isFetching: false,
    items: [],
    errorMessage: ""
  }
}

export default Dragons;