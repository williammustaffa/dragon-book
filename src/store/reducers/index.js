import { combineReducers } from "redux";
import { connectRouter } from "connected-react-router";
import dragons from "./dragons";
import dragon from "./dragon";
import user from "./user";

const createRootReducer = (history) => combineReducers({
  router: connectRouter(history),
  dragons,
  dragon,
  user,
});

export default createRootReducer;