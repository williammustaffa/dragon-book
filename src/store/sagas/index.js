import { all } from "redux-saga/effects";

// User
import { watchUserLoginAsync } from "./user/userLogin";
import { watchUserLogoutAsync } from "./user/userLogout";
import { watchUserCheckSessionAsync } from "./user/userCheckSession";
import { watchCreateProfileAsync } from "./user/createProfile";

// Dragons
import { watchFetchDragons } from "./dragons/fetchDragons";
import { watchFetchDragon } from "./dragon/fetchDragon";

/**
 * Root saga provided to redux middleware
 */
function* root() {
  const watchers = [];

  // User watchers
  watchers.push(watchUserLoginAsync());
  watchers.push(watchUserLogoutAsync());
  watchers.push(watchUserCheckSessionAsync());
  watchers.push(watchCreateProfileAsync());

  // Dragons watchers
  watchers.push(watchFetchDragon());
  watchers.push(watchFetchDragons());

  yield all(watchers);
}

export default root;