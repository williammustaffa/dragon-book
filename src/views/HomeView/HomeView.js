import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { fetchDragons } from "store/actions";
import { Grid } from "semantic-ui-react";
import { DragonCard } from "components/DragonCard";

function HomeView() {
  const dispatch = useDispatch();

  const { dragons } = useSelector(function (state) {
    return {
      dragons: state.dragons.items,
    }
  });

  useEffect(function () {
    dispatch(fetchDragons());
  }, [dispatch]);

  const renderDragonCard = dragon => (
    <Grid.Column key={dragon.id} stretched mobile={16} tablet={8} computer={4}>
      <DragonCard dragon={dragon} />
    </Grid.Column>
  );

  return (
    <Grid>
      {dragons.map(renderDragonCard)}
    </Grid>
  );
}

export default HomeView;