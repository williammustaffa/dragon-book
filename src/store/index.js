import { applyMiddleware, compose, createStore } from "redux";
import { routerMiddleware } from "connected-react-router";
import createRootReducer from "./reducers";
import sagas from "./sagas";
import history from "./history";
import createSagaMiddleware from "redux-saga"
import logger from "redux-logger";
import configureStore from "redux-mock-store";

/**
 * Generates a redux store properly
 * applying reducers, states and middlewares
 *
 * @param {*} preloadedState 
 */
function generateStore(preloadedState) {
  const sagaMiddleware = createSagaMiddleware();

  const store = createStore(
    createRootReducer(history),
    preloadedState,
    compose(
      applyMiddleware(
        routerMiddleware(history),
        sagaMiddleware,
        logger,
      ),
    ),
  );

  // Run redux saga
  sagaMiddleware.run(sagas);

  return store;
}

/**
 * Mock store and return a mocked state
 * @param {*} preloadedState 
 */
function mockStore(preloadedState) {
  const middlewares = [];
  const generateMockStore = configureStore(middlewares);

  return generateMockStore(preloadedState);
}

export {
  generateStore,
  mockStore,
  history,
};