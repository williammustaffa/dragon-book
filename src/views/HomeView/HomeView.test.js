import React from 'react';
import HomeView from './HomeView';
import Dragons from 'store/models/Dragons';
import { Provider } from 'react-redux';
import { mockStore } from "store";
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

test('Test home component', () => {
  // Mock store
  const store = mockStore({
    dragons: Dragons.state
  });

  const { asFragment } = render(
    <Provider store={store}>
      <HomeView />
    </Provider>
  );

  // Run tests
  expect(asFragment()).toMatchSnapshot();
});