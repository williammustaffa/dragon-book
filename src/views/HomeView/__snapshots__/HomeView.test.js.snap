// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test home component 1`] = `
<DocumentFragment>
  <div
    class="ui grid"
  />
</DocumentFragment>
`;
